src/ClientData/CaptureData.cpp

```diff
@@ -251,4 +252,31 @@ void CaptureData::UpdateTimerDurations() {
   return result;
 }
 
+[[nodiscard]] std::optional<ThreadStateSliceInfo>
+CaptureData::FindThreadStateSliceInfoFromTimestamp(int64_t thread_id, uint64_t timestamp) const {
+  absl::MutexLock lock{&thread_state_slices_mutex_};
+  if (!thread_state_slices_.contains(thread_id)) {
+    return std::nullopt;
+  }
+
+  const std::vector<ThreadStateSliceInfo>& thread_state_bar = thread_state_slices_.at(thread_id);
+  auto slice =
+      std::upper_bound(  // This should return the slice directly in front of the slice that we want
+          thread_state_bar.begin(), thread_state_bar.end(), timestamp,
+          [](uint64_t b,
+             const ThreadStateSliceInfo& a) -> bool {  // compare based on beginning timestamps
+            return b < a.begin_timestamp_ns();
+          });
+
+  if (slice == thread_state_bar.begin()) {
+    return std::nullopt;
+  }
```

Comment:
I just realized that `slice` might also be `thread_state_bar.end()`. For example when `thread_state_bar` is empty.

So can you also check for the end iterator and return nullopt in that case? And maybe also add a test for the case when there are no ThreadStateSlices in the vector?